src/pybind/mgr/dashboard/controllers/user.py

```diff
@@ -12,6 +13,37 @@
 from ..services.auth import JwtManager
 
 
+# minimum password complexity rules
+def check_password_complexity(password, username):
+    """
+    Suggested rules for password complexity
+    - at least 6 chars in length
+    - must not be the same as the user account name
+    - consist of characters from the following groups
+    - alphabetic a-z, A-Z
+    - numbers 0-9
+    - special chars: !_@
+    - must use at least 1 special char
+    """
+    if password == username:
+        raise DashboardException(msg='Password is the same as the username.\
+                                      It has to be different',
+                                 code='password-the-same-as-username',
+                                 component='Update/Create user')
+
+    password_min_complex = re.compile('^(?=\\S{6,20}$)(?=.*[!_@])(?=.*[a-z])'
+                                      '(?=.*[A-Z])(?=.*[0-9])')
```

Comment:
While the regex approach looks well crafted, in this case I'm not sure if it's the best one, as:
- It's less understandable and hence harder to debug/extend,
- It does not allow to report which specific rule has failed.

I'd suggest instead using a multi-step check:
```python
if not test_01(password):
   ...
elif not test_02(password):
   ...
else:
  # password valid
```

Reply:
During our F2F meeting, the usage of the [password strength](https://pypi.org/project/password-strength/) Python module [was discussed](https://pad.ceph.com/p/dashboard-f2f-2019-password-management) - maybe that would help to test for some of these requirements?